In [1]:
import os
os.listdir("../Data")

FileNotFoundError: [WinError 3] Sistem belirtilen yolu bulamıyor: '../Data'

In [2]:
import os
os.getcwd()

'C:\\Users\\Recep Sadun\\OneDrive\\Masaüstü\\GitHup için ilk klasör'

In [3]:
import os
os.listdir("Data")

['spotify-2023.csv']

In [6]:
import pandas as pd

df = pd.read_csv("Data/spotify-2023.csv", encoding="latin1")
df.head()

track_name    artist(s)_name  artist_count  \
0  Seven (feat. Latto) (Explicit Ver.)  Latto, Jung Kook             2   
1                                 LALA       Myke Towers             1   
2                              vampire    Olivia Rodrigo             1   
3                         Cruel Summer      Taylor Swift             1   
4                       WHERE SHE GOES         Bad Bunny             1   

   released_year  released_month  released_day  in_spotify_playlists  \
0           2023               7            14                   553   
1           2023               3            23                  1474   
2           2023               6            30                  1397   
3           2019               8            23                  7858   
4           2023               5            18                  3133   

   in_spotify_charts    streams  in_apple_playlists  ...  bpm key   mode  \
0                147  141381703                  43  ...  125   B  Major   
1                 48  133716286                  48  ...   92  C#  Major   
2                113  140003974                  94  ...  138   F  Major   
3                100  800840817                 116  ...  170   A  Major   
4                 50  303236322                  84  ...  144   A  Minor   

  danceability_%  valence_% energy_% acousticness_%  instrumentalness_%  \
0             80         89       83             31                   0   
1             71         61       74              7                   0   
2             51         32       53             17                   0   
3             55         58       72             11                   0   
4             65         23       80             14                  63   

   liveness_%  speechiness_%  
0           8              4  
1          10              4  
2          31              6  
3          11             15  
4          11              6  

[5 rows x 24 columns]

In [10]:
df.columns

Index(['track_name', 'artist(s)_name', 'artist_count', 'released_year',
       'released_month', 'released_day', 'in_spotify_playlists',
       'in_spotify_charts', 'streams', 'in_apple_playlists', 'in_apple_charts',
       'in_deezer_playlists', 'in_deezer_charts', 'in_shazam_charts', 'bpm',
       'key', 'mode', 'danceability_%', 'valence_%', 'energy_%',
       'acousticness_%', 'instrumentalness_%', 'liveness_%', 'speechiness_%'],
      dtype='object')

In [11]:
df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   track_name            953 non-null    object
 1   artist(s)_name        953 non-null    object
 2   artist_count          953 non-null    int64 
 3   released_year         953 non-null    int64 
 4   released_month        953 non-null    int64 
 5   released_day          953 non-null    int64 
 6   in_spotify_playlists  953 non-null    int64 
 7   in_spotify_charts     953 non-null    int64 
 8   streams               953 non-null    object
 9   in_apple_playlists    953 non-null    int64 
 10  in_apple_charts       953 non-null    int64 
 11  in_deezer_playlists   953 non-null    object
 12  in_deezer_charts      953 non-null    int64 
 13  in_shazam_charts      903 non-null    object
 14  bpm                   953 non-null    int64 
 15  key                   858 non-null    ob

### Veri Seti Genel Özeti

Bu veri seti 2023 yılında Spotify’da en çok dinlenen şarkılara ait 
**953 satır ve 24 sütundan** oluşmaktadır.

- **Sayısal değişkenler:** streams, in_spotify_playlists, in_spotify_charts ve 
  müzik özellikleri (danceability, energy, valence vb.)
- **Kategorik değişkenler:** track_name, artist(s)_name, key ve mode
- Çoğu kolonda eksik veri bulunmamakla birlikte, `key` ve `in_shazam_charts`
  değişkenlerinde kısmi eksiklikler mevcuttur.

Veri, hem **popülerlik (stream, playlist, chart)** hem de 
**müzikal özellikler** açısından analiz yapmaya uygundur.


In [12]:
df.isnull().sum().sort_values(ascending=False)

key                     95
in_shazam_charts        50
track_name               0
artist(s)_name           0
released_month           0
released_day             0
artist_count             0
released_year            0
in_spotify_charts        0
in_spotify_playlists     0
streams                  0
in_apple_playlists       0
in_deezer_playlists      0
in_apple_charts          0
in_deezer_charts         0
bpm                      0
mode                     0
danceability_%           0
valence_%                0
energy_%                 0
acousticness_%           0
instrumentalness_%       0
liveness_%               0
speechiness_%            0
dtype: int64

## Eksik Veri Analizi

Veri setinde en fazla eksik değere sahip kolonlar:
- `key` (95 eksik)
- `in_shazam_charts` (50 eksik)

Eksik değer oranları, toplam veri boyutuna kıyasla düşüktür.
Bu nedenle analizlerin büyük kısmı bu eksikliklerden etkilenmemektedir.

Müzikal özellikler ve Spotify/Apple metrikleri eksiksiz olduğu için,
popülerlik ve müzik karakteristiklerine yönelik analizler güvenle yapılabilir.

In [13]:
df.sort_values("streams", ascending=False).head(10)[
    ["track_name", "artist(s)_name", "streams"]
]


track_name                 artist(s)_name  \
574    Love Grows (Where My Rosemary Goes)              Edison Lighthouse   
33                               Anti-Hero                   Taylor Swift   
625                                 Arcade                Duncan Laurence   
253                          Glimpse of Us                           Joji   
455                         Seek & Destroy                            SZA   
98                      Summertime Sadness                   Lana Del Rey   
891  Come Back Home - From "Purple Hearts"                   Sofia Carson   
427                      Where Are You Now  Lost Frequencies, Calum Scott   
322                          I Love You So                    The Walters   
130                              Queencard                       (G)I-DLE   

                                               streams  
574  BPM110KeyAModeMajorDanceability53Valence75Ener...  
33                                           999748277  
625                                          991336132  
253                                          988515741  
455                                           98709329  
98                                           983637508  
891                                           97610446  
427                                          972509632  
322                                          972164968  
130                                           96273746

2023 yılında en yüksek dinlenme sayısına sahip şarkılar, belirli sanatçılar etrafında yoğunlaşmaktadır.
Bu durum, popülerliğin sınırlı sayıda sanatçı üzerinde toplandığını göstermektedir

In [16]:
df.groupby("artist(s)_name")["streams"].sum() \
  .sort_values(ascending=False) \
  .head(10)

artist(s)_name
Edison Lighthouse                BPM110KeyAModeMajorDanceability53Valence75Ener...
Duncan Laurence                                                          991336132
Joji                                                  9885157412463900681122364376
Sofia Carson                                                              97610446
Lost Frequencies, Calum Scott                                            972509632
The Walters                                                              972164968
(G)I-DLE                                                         96273746170709584
The Weeknd, Future                                                        96180277
IU, Agust D                                                               95816024
Sean Paul, Dua Lipa                                                      956865266
Name: streams, dtype: object

## Veri Tipi Düzeltmeleri

In [17]:
df["streams"] = pd.to_numeric(df["streams"], errors="coerce")

EDA sırasında streams değişkeninin metin (object) tipinde olduğu fark edilmiştir.
Analizlerin doğru yapılabilmesi için bu değişken sayısal tipe dönüştürülmüştür.

## Veri Tipi Düzeltmesi Sonrası Güncellenen Analizler

In [18]:
df.sort_values("streams", ascending=False).head(10)[
    ["track_name", "artist(s)_name", "streams"]
]


track_name  \
55                                 Blinding Lights   
179                                   Shape of You   
86                               Someone You Loved   
620                                   Dance Monkey   
41   Sunflower - Spider-Man: Into the Spider-Verse   
162                                      One Dance   
84                       STAY (with Justin Bieber)   
140                                       Believer   
725                                         Closer   
48                                         Starboy   

                   artist(s)_name       streams  
55                     The Weeknd  3.703895e+09  
179                    Ed Sheeran  3.562544e+09  
86                  Lewis Capaldi  2.887242e+09  
620                   Tones and I  2.864792e+09  
41          Post Malone, Swae Lee  2.808097e+09  
162           Drake, WizKid, Kyla  2.713922e+09  
84   Justin Bieber, The Kid Laroi  2.665344e+09  
140               Imagine Dragons  2.594040e+09  
725      The Chainsmokers, Halsey  2.591224e+09  
48          The Weeknd, Daft Punk  2.565530e+09

In [19]:
df.groupby("artist(s)_name")["streams"].sum() \
  .sort_values(ascending=False) \
  .head(10)

artist(s)_name
The Weeknd         1.418555e+10
Taylor Swift       1.405366e+10
Ed Sheeran         1.390895e+10
Harry Styles       1.160865e+10
Bad Bunny          9.997800e+09
Olivia Rodrigo     7.442149e+09
Eminem             6.183806e+09
Bruno Mars         5.846921e+09
Arctic Monkeys     5.569807e+09
Imagine Dragons    5.272485e+09
Name: streams, dtype: float64

In [20]:
df.sort_values("streams", ascending=False).head(10)[
    ["track_name", "artist(s)_name", "streams"]
]

track_name  \
55                                 Blinding Lights   
179                                   Shape of You   
86                               Someone You Loved   
620                                   Dance Monkey   
41   Sunflower - Spider-Man: Into the Spider-Verse   
162                                      One Dance   
84                       STAY (with Justin Bieber)   
140                                       Believer   
725                                         Closer   
48                                         Starboy   

                   artist(s)_name       streams  
55                     The Weeknd  3.703895e+09  
179                    Ed Sheeran  3.562544e+09  
86                  Lewis Capaldi  2.887242e+09  
620                   Tones and I  2.864792e+09  
41          Post Malone, Swae Lee  2.808097e+09  
162           Drake, WizKid, Kyla  2.713922e+09  
84   Justin Bieber, The Kid Laroi  2.665344e+09  
140               Imagine Dragons  2.594040e+09  
725      The Chainsmokers, Halsey  2.591224e+09  
48          The Weeknd, Daft Punk  2.565530e+09

Streams değişkeni sayısal tipe dönüştürüldükten sonra, en çok dinlenen şarkılar doğru şekilde listelenmiştir.

In [21]:
df.groupby("in_spotify_playlists")["streams"].mean() \
  .sort_index()

in_spotify_playlists
31       3.841196e+07
34       6.821699e+07
58       3.905856e+07
67       4.795638e+07
77       2.956222e+07
             ...     
44927    1.479115e+09
49991    1.690193e+09
50887    1.970673e+09
51979    1.806618e+09
52898    9.338156e+08
Name: streams, Length: 879, dtype: float64

Spotify playlist sayısı ile ortalama stream sayısı arasında
düzgün ve sürekli bir artış ilişkisi bulunmamaktadır.
Bununla birlikte, yüksek playlist sayısına sahip şarkıların
ortalama stream değerlerinin genel olarak daha yüksek olduğu
gözlemlenmektedir.


In [22]:
df["playlist_group"] = pd.cut(
    df["in_spotify_playlists"],
    bins=[-1, 0, 10, 50, 200, df["in_spotify_playlists"].max()],
    labels=["0", "1–10", "11–50", "51–200", "200+"]
)

df.groupby("playlist_group")["streams"].mean()


C:\Users\Recep Sadun\AppData\Local\Temp\ipykernel_35844\3798998256.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby("playlist_group")["streams"].mean()


playlist_group
0                  NaN
1–10               NaN
11–50     5.331447e+07
51–200    1.021860e+08
200+      5.262676e+08
Name: streams, dtype: float64

In [23]:
df[df["playlist_group"] == "0"][["track_name", "streams"]].head()
df[df["playlist_group"] == "1–10"][["track_name", "streams"]].head()


Empty DataFrame
Columns: [track_name, streams]
Index: []

Playlist gruplarına göre yapılan analizde,
0 ve 1–10 aralığında playlist sayısına sahip
şarkılar için yeterli gözlem bulunmadığı görülmektedir.

Buna karşın 11–50 ve üzeri playlist gruplarında
ortalama stream değerleri belirgin şekilde artmaktadır.
Bu durum, yüksek playlist görünürlüğünün
şarkı popülerliği ile ilişkili olduğunu göstermektedir.

"0 ve 1–10 playlist gruplarında yer alan şarkılara
veri setinde rastlanmamıştır.
Bu nedenle bu gruplar için ortalama stream analizi
yapılamamıştır."
